From 7ad882f7f444816cd859ae1d60773ac35fbad395 Mon Sep 17 00:00:00 2001
From: Carlos Munoz <cmunoz@caviumnetworks.com>
Date: Wed, 6 Jan 2016 13:05:55 -0800
Subject: [PATCH 08/13] netdev: octeon-pow: Account for size of wqe to
 calculate start of fpa buffer.

Signed-off-by: Carlos Munoz <cmunoz@caviumnetworks.com>
[Original patch taken from patch set for OCTEON SDK 3.1.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/net/ethernet/octeon/octeon-pow-ethernet.c |   18 +++++++++++-------
 1 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/octeon/octeon-pow-ethernet.c b/drivers/net/ethernet/octeon/octeon-pow-ethernet.c
index d043917..b542112 100644
--- a/drivers/net/ethernet/octeon/octeon-pow-ethernet.c
+++ b/drivers/net/ethernet/octeon/octeon-pow-ethernet.c
@@ -190,18 +190,22 @@ octeon_get_next_packet_ptr(union octeon_packet_ptr packet_ptr)
 
 static int octeon_pow_free_work(cvmx_wqe_t *work)
 {
-	union octeon_packet_ptr	segment_ptr;
-	union octeon_packet_ptr	next_ptr;
+	union octeon_packet_ptr	packet_ptr;
 	int			segments;
+	void			*buffer_ptr;
 
 	segments = cvmx_wqe_get_bufs(work);
 
-	segment_ptr = octeon_get_first_packet_ptr(work);
+	packet_ptr = octeon_get_first_packet_ptr(work);
+	if (octeon_has_feature(OCTEON_FEATURE_PKI))
+		buffer_ptr = (void *)((ulong)work - 128);
+	else
+		buffer_ptr = get_buffer_ptr(packet_ptr);
+
 	while (segments--) {
-		next_ptr = octeon_get_next_packet_ptr(segment_ptr);
-		cvmx_fpa_free(get_buffer_ptr(segment_ptr),
-			      cvmx_wqe_get_aura(work), 0);
-		segment_ptr = next_ptr;
+		packet_ptr = octeon_get_next_packet_ptr(packet_ptr);
+		cvmx_fpa_free(buffer_ptr, cvmx_wqe_get_aura(work), 0);
+		buffer_ptr = get_buffer_ptr(packet_ptr);
 	}
 
 	if (!octeon_has_feature(OCTEON_FEATURE_PKI))
-- 
1.7.5.4

